crates/red_knot_test/src/lib.rs

```diff
@@ -102,24 +101,9 @@ fn run_test(db: &mut db::Db, test: &parser::MarkdownTest) -> Result<(), Failures
     let failures: Failures = test_files
         .into_iter()
         .filter_map(|test_file| {
-            let parsed = parsed_module(db, test_file.file);
-
-            let mut diagnostics: Vec<Box<_>> = parsed
-                .errors()
-                .iter()
-                .cloned()
-                .map(|error| {
-                    let diagnostic: Box<dyn Diagnostic> =
-                        Box::new(ParseDiagnostic::new(test_file.file, error));
-                    diagnostic
-                })
-                .collect();
-
-            let type_diagnostics = check_types(db, test_file.file);
-            diagnostics.extend(type_diagnostics.into_iter().map(|diagnostic| {
-                let diagnostic: Box<dyn Diagnostic> = Box::new((*diagnostic).clone());
-                diagnostic
-            }));
+            let mut diagnostics = check_types::accumulated::<CompileDiagnostic>(db, test_file.file);
+            // Filter out diagnostics that are not related to the current file.
```

Comment:
Only tangentially related: do we have a feeling for how many unrelated-file diagnostics we generate typically? If this would be a large fraction of all diagnostics, we should probably detect that earlier?

Reply:
This could be any number of diagnostics and salsa doesn't provide a way to detect this earlier. 

Salsa accumulators work by traversing the entire dependency tree of a query. In our case, this is `check_file`. `check_file` can branch out into arbitrary files when resolving imports (and checking those files in turn). The only way for us to truncate this earlier is by not using accumulators because salsa doesn't know about the file boundaries. 

I don't think this is very terrible because the CLI uses `check_workspace` to get *all* diagnostics. The LSP uses `check_file` today but I don't think it should. The wasm playground uses `check_file`... this is probably fine